### ISSUE TITLE
Allow user customization of release notes template

### ISSUE THREAD

**Issue by jmgate** [python-semantic-release/python-semantic-release#660]
## Description

It would be great if `semantic-release version` allowed the user to customize how the release notes posted to GitHub, etc., get generated.

## Use cases

The default template used doesn't work well with commits of the form
```
type: Commit summary

* Foo
* Bar
* Baz
```
because they wind up rendering as
* type: Commit summary (SHA1)
* Foo
* Bar
* Baz

instead of
* type: Commit summary (SHA1)
  * Foo
  * Bar
  * Baz

It'd also be great to just allow flexibility so project maintainers can tailor their release notes to their heart's content.

## Possible implementation

This is just a guess:
* Add a `template_file` argument to `render_release_notes()` in `semantic_release/cli/common.py`.
  * Default it to `None`.
  * If it's `None`, set it to `files("semantic_release").joinpath("data/templates/release_notes.md.j2")`, as is currently done.
  * Read it into the `release_template` string.
* Add a `@click.option` for `--release-template` to `version()` in `semantic_release/cli/commands/version.py`.
  * Default it to `None`.
  * Pass it on to `render_release_notes()`.

**Comment by bryant-finney** [comment]
@jmgate thanks for creating this issue; I also need this functionality. After looking into the source code a bit, I'd like to tweak this possible implementation to use the following implementation plan.

### Implementation Plan

1. refactor: [`render_release_notes()`](https://github.com/python-semantic-release/python-semantic-release/blob/78f638757ebf482f9bc110fca7cdb873265b0b7a/semantic_release/cli/common.py#L19) → `render_default_release_notes()`
2. surround [semantic_release/cli/commands/changelog.py#91-93](https://github.com/python-semantic-release/python-semantic-release/blob/78f638757ebf482f9bc110fca7cdb873265b0b7a/semantic_release/cli/commands/changelog.py#L91-L93) and [semantic_release/cli/commands/version.py#L530-L534](https://github.com/python-semantic-release/python-semantic-release/blob/78f638757ebf482f9bc110fca7cdb873265b0b7a/semantic_release/cli/commands/version.py#L530-L534) with logic to test for existence of `template_dir / "release_notes.md.j2"` 
   - if it exists, render the release notes from this template
   - otherwise, use `render_default_release_notes()` 

@bernardcooke53 any thoughts or suggestions for this implementation plan?

**Comment by jmgate** [comment]
Sounds good to me.

**Comment by bryant-finney** [comment]
### Update 0

I finished implementing [two (failing) acceptance tests](https://github.com/bryant-finney/python-semantic-release/actions/runs/6565874430/job/17835362801#step:6:6219):

1. bryant-finney/python-semantic-release@acde492f6d89de9f36c6807f0336d0bc2b78ee43 exercises the command `semantic-release changelog --post-to-release-tag v0.2.0-rc.1`
2. bryant-finney/python-semantic-release@8e18b98013f08ff082f882695abbf164852f3f8f exercises the command `semantic-release version --skip-build --vcs-release`

In both tests, [`requests.Session.post()`](https://github.com/psf/requests/blob/main/src/requests/sessions.py#L626-L637) is patched to capture the `POST` request that creates the new release. The release notes are sent with the request body and are extracted for comparison.

Locally, I was able to get both of these tests to pass by changing [semantic_release/data/templates/release_notes.md.j2](https://github.com/python-semantic-release/python-semantic-release/blob/78f638757ebf482f9bc110fca7cdb873265b0b7a/semantic_release/data/templates/release_notes.md.j2) to...

```markdown
## What's Changed
{% for type_, commits in release["elements"] | dictsort %}
### {{ type_ | capitalize }}
{%- if type_ != "unknown" %}
{% for commit in commits %}
* {{ commit.commit.summary.rstrip() }} ([`{{ commit.short_hash }}`]({{ commit.hexsha | commit_hash_url }}))
{%- endfor %}{% endif %}{% endfor %}
```

@jmgate , please let me know if you see any issues with these tests 🙏

**Comment by jmgate** [comment]
Looks good to me.  Very thorough, and I appreciate seeing the failing tests before the feature addition.  Also learned a thing or two about complex mocking and type checking that I can bring back to my own projects, so thanks for that.

**Comment by jmgate** [comment]
Implemented in #736.

Using [this template](https://github.com/sandialabs/reverse_argparse/blob/master/templates/.release_notes.md.j2) resulted in [these release notes](https://github.com/sandialabs/reverse_argparse/releases/tag/v1.0.5).

![Screenshot 2023-10-24 at 10 32 29 AM](https://github.com/python-semantic-release/python-semantic-release/assets/20327215/83dfc21e-0bd4-4d6f-85c1-e6d351d4322e)

Thanks @bryant-finney&mdash;this feature is awesome! 🎉
